### ISSUE TITLE
could you update the good way of D2Dnet in yolo?

### ISSUE THREAD

**Issue by sankexin** [ultralytics/yolov5#1014]
## ❔Question
could update the good way of D2Dnet in yolo?

## Additional context
D2Det: https://github.com/JialeCao001/D2Det
![2](https://user-images.githubusercontent.com/33318353/93870966-874cb500-fd00-11ea-81ac-f96db67334b7.jpg)
![1](https://user-images.githubusercontent.com/33318353/93870979-8c116900-fd00-11ea-822c-05d242244738.jpg)

**Comment by sankexin** [comment]
> Hello @sankexin, thank you for your interest in our work! Please visit our [Custom Training Tutorial](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data) to get started, and see our [Jupyter Notebook](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) [![Open In Colab](https://camo.githubusercontent.com/52feade06f2fecbf006889a904d221e6a730c194/68747470733a2f2f636f6c61622e72657365617263682e676f6f676c652e636f6d2f6173736574732f636f6c61622d62616467652e737667)](https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb), [Docker Image](https://hub.docker.com/r/ultralytics/yolov5), and [Google Cloud Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) for example environments.
> 
> If this is a bug report, please provide screenshots and **minimum viable code to reproduce your issue**, otherwise we can not help you.
> 
> If this is a custom model or data training question, please note Ultralytics does **not** provide free personal support. As a leader in vision ML and AI, we do offer professional consulting, from simple expert advice up to delivery of fully customized, end-to-end production solutions for our clients, such as:
> 
> * **Cloud-based AI** systems operating on **hundreds of HD video streams in realtime.**
> * **Edge AI** integrated into custom iOS and Android apps for realtime **30 FPS video inference.**
> * **Custom data training**, hyperparameter evolution, and model exportation to any destination.
> 
> For more information please visit https://www.ultralytics.com.

yolov5 is as good as yolov4, maybe this way will make it much more better than yolov4, so I wish it is a good idea.

**Comment by glenn-jocher** [comment]
@sankexin I briefly reviewed the repo. It seems D2D is a two-stage method, which is naturally not aligned with the single-stage detection premise behind all versions of YOLO. Are there specific parts of the D2D implementation that you think might be adoptable here?

**Comment by sankexin** [comment]
> @sankexin I briefly reviewed the repo. It seems D2D is a two-stage method, which is naturally not aligned with the single-stage detection premise behind all versions of YOLO. Are there specific parts of the D2D implementation that you think might be adoptable here?

No need to consider the tools of the first stage. Yolo can take advantage of the tools of the second stage of this paper after RPN，such as, box location for increasing AP 、binary overlap prediction for Reducing background false positives 、adaptive weighting to obtain discriminative features for Reducing false positives of the other class.

**Comment by glenn-jocher** [comment]
@sankexin ah I see. Yes, definitely selecting the best elements from various sources for integration is a good recipe for success. I don't have much free time to investigate these myself, but if find any positive results in your own research please let us know.

Reducing FPs is useful for real life scenarios, though this will often also result in mAP reductions on validations sets. We have an option in NMS that requires detection redundancy to reduce FPs, though this requires computation of IoUs outside of the pytorch NMS function, slowing NMS.
https://github.com/ultralytics/yolov5/blob/7220cee1d1dc1f14003dbf8d633bbb76c547000c/utils/general.py#L660-L661

**Comment by sankexin** [comment]
> @sankexin ah I see. Yes, definitely selecting the best elements from various sources for integration is a good recipe for success. I don't have much free time to investigate these myself, but if find any positive results in your own research please let us know.
> 
> Reducing FPs is useful for real life scenarios, though this will often also result in mAP reductions on validations sets. We have an option in NMS that requires detection redundancy to reduce FPs, though this requires computation of IoUs outside of the pytorch NMS function, slowing NMS.
> https://github.com/ultralytics/yolov5/blob/7220cee1d1dc1f14003dbf8d633bbb76c547000c/utils/general.py#L660-L661

Thanks for your reply. I'll try your method. I have tried NMS、 focalloss、GHM and aploss on Yolov3 before, but the effect of reducing FPs is very little. The method of this paper has given me new hope, but there are some code that I con't realize at the moment.

**Comment by wzj5133329** [comment]
@sankexin hi, have you used d2det on Yolo? especially dense local regression...
